third_party/pcsc-lite/naclport/server_clients_management/src/admin-policy-service.js

```diff
@@ -0,0 +1,207 @@
+goog.require('goog.Disposable');
+goog.require('goog.log');
+goog.require('goog.log.Logger');
+goog.require('goog.object');
+goog.require('goog.messaging.AbstractChannel');
+
+goog.scope(function() {
+
+/**
+ * Service name that is used for the message to the PC/SC server that the admin
+ * policy has been updated.
+ *
+ * The message data will contain the data of the new admin policy.
+ */
+const UPDATE_ADMIN_POLICY_SERVER_MESSAGE_SERVICE_NAME = 'update_admin_policy';
+
+const GSC = GoogleSmartCard;
+const DeferredProcessor = GSC.DeferredProcessor;
+const ReadinessTracker = GSC.PcscLiteServerClientsManagement.ReadinessTracker;
+
+/**
+ * This class provides the C++ layer with the AdminPolicy obtained from the
+ * managed storage. Upon initialization, it reads the current value from
+ * storage and sends a message containing the policy data.
+ * @param {!goog.messaging.AbstractChannel} serverMessageChannel Message channel
+ * to the PC/SC server into which the admin policy will be forwarded.
+ * @param {!ReadinessTracker} serverReadinessTracker Tracker of the PC/SC server
+ * that allows to delay forwarding of the messages to the server until it is
+ * ready to receive them.
+ * @constructor
+ * @extends goog.Disposable
+ */
+GSC.PcscLiteServerClientsManagement.AdminPolicyService = function(
+    serverMessageChannel, serverReadinessTracker) {
+  AdminPolicyService.base(this, 'constructor');
+
+  /**
+   * @type {!goog.log.Logger}
+   * @const
+   */
+  this.logger = GSC.Logging.getScopedLogger(
+      'PcscLiteServerClientsManagement.AdminPolicyService');
+
+  /** @private */
+  this.serverMessageChannel_ = serverMessageChannel;
+
+  /** @private */
+  this.serverReadinessTracker_ = serverReadinessTracker;
+
+  // The policy update messages are deferred until the server readiness is
+  // established.
+  /** @private */
+  this.deferredProcessor_ =
+      new DeferredProcessor(this.serverReadinessTracker_.promise);
+
+  this.addChannelDisposedListener_();
+
+  /**
+   * @type {!Object}
+   * @private
+   */
+  this.adminPolicy_ = {};
+
+  this.loadManagedStorage_();
+  this.listenForStorageChanging_();
+
+  goog.log.fine(this.logger, 'Initialized');
+};
+
+const AdminPolicyService =
+    GSC.PcscLiteServerClientsManagement.AdminPolicyService;
+
+goog.inherits(AdminPolicyService, goog.Disposable);
+
+/** @override */
+AdminPolicyService.prototype.disposeInternal = function() {
+  this.deferredProcessor_.dispose();
+  this.deferredProcessor_ = null;
+
+  this.serverReadinessTracker_ = null;
+
+  this.serverMessageChannel_ = null;
+
+  goog.log.fine(this.logger, 'Disposed');
+
+  AdminPolicyService.base(this, 'disposeInternal');
+};
+
+/** @private */
+AdminPolicyService.prototype.loadManagedStorage_ = function() {
+  chrome.storage.managed.get(this.managedStorageLoadedCallback_.bind(this));
+};
+
+/**
+ * @param {!Object} items
+ * @private
+ */
+AdminPolicyService.prototype.managedStorageLoadedCallback_ = function(items) {
+  goog.log.fine(
```

Comment:
nit: The "fine" log level is off in Release builds IIRC. As I think the logs are useful, let's use something like the INFO log level.